### PR TITLE
Let the CTC prefix scorer run on the CPU while the model decodes on an accelerator

### PR DESCRIPTION
**Pull request by sw005320**
Follow-up to #6599, independent of #6610.

## What did you change?

`CTCPrefixScorer` gains a `scoring_device`. The CTC posteriors are moved there once per utterance in `batch_init_state`, the prefix-score states stay there, and `batch_score_partial` hands the scores back on the encoder's device, so the beam search sees no difference. `batch_select_state` and `extend_prob` (streaming) move their inputs the same way. The default, `None`, follows the encoder as before.

`Speech2Text` takes `ctc_scoring_device`, exposed as `--ctc_scoring_device`:

- `same`: scores stay with the model.
- `auto` (default): the CPU when the model is on an accelerator (CUDA or MPS), the model's device for a CPU model. Both accelerators were measured to benefit, see below.
- anything else names a device, e.g. `cpu`.

This changes where existing GPU recipes compute the CTC prefix score; the transcripts are unchanged up to floating point (identical in every measurement below), and `--ctc_scoring_device same` restores the previous placement. `inference()` takes the new argument with a default, so programmatic callers keep working. Other inference bins are untouched; threading the option through `asr_inference_streaming.py` and `s2t_inference.py` is a follow-up if this proves useful.

## Why did you make this change?

The forward recursion of the CTC prefix score walks the encoder frames one by one, a few tiny kernels per frame. That is cheap on a CPU and dear on an accelerator with a high launch cost. Profiling one `batch_size=1` decoding step of the LibriSpeech E-Branchformer, beam 10, CTC 0.3:

| | MPS | CPU |
|---|---:|---:|
| whole step | 114 ms | 136 ms |
| CTC prefix score | 48 ms | 4 ms |
| decoder | 36 ms | 42 ms |
| post_process + make_batch | 15 ms | 0.5 ms |

The same recursion, the same work, twelve times the wall clock on the GPU: the step there is bound by kernel launches, not by arithmetic. So the recursion goes where launches are cheap and the decoder, whose matmuls do benefit from the GPU, stays where it is. The per-step traffic is the prefix tokens and pre-beam candidates in one direction and an `(n_hyp, vocab)` score matrix in the other, a few hundred kilobytes.

**End to end**, trained models on real audio, MPS, beam 10, CTC 0.3, exact CTC (no window). `same` against `auto`, interleaved, two repeats each; transcripts identical in every run.

| | CTC on MPS | CTC on CPU | |
|---|---:|---:|---:|
| LibriSpeech test-clean, 16 utts, batch_size 1 | 95.0 / 100.7 s | 44.7 / 52.6 s | 2.01x |
| LibriSpeech test-clean, 16 utts, batch_size 4 | 100.2 / 107.9 s | 52.9 / 60.7 s | 1.83x |
| AISHELL-1 test, 50 utts, batch_size 1 | 93.8 / 112.6 s | 45.2 / 68.1 s | 1.82x |
| AISHELL-1 test, 50 utts, batch_size 4 | 46.8 / 61.4 s | 29.4 / 36.6 s | 1.64x |

With the window of #6610 on (margin 100) the recursion is shorter and the gain smaller; the prototype measured 32 s to 17 s over 8 LibriSpeech utterances in that configuration.

**CUDA**, Colab T4 with two slow CPU cores, same model, exact CTC, 16 test-clean utterances, two interleaved repeats, transcripts identical (16/16) in every run:

| | CTC on CUDA | CTC on CPU | |
|---|---:|---:|---:|
| batch_size 1 | 38.0 / 37.1 s | 25.8 / 26.2 s | 1.45x |
| batch_size 4 | 31.0 / 32.0 s | 31.4 / 32.8 s | 0.98x |
| CTC per step, batch_size 1 | 50 ms | 27 ms | |
| CTC per step, batch_size 4 | 78 ms | 73 ms | |

The GPU side is launch-bound (50 ms for the same recursion that takes 4 ms on the Mac's CPU), so it will not improve with a faster GPU, while the CPU side on a server is several times faster than Colab's two vCPUs. `auto` therefore picks the CPU on CUDA as well. The batch_size 4 rows there were measured with a timing wrapper that hid `batch_score`'s signature, so the padding mask was not applied; that changes what is attended to, not the amount of work, and it is why #6628 exists.

**A pitfall found on the way.** A prototype copied the scores back with `non_blocking=True` and produced a garbage transcript once out of eight: on MPS the copy from unpinned host memory can complete after the source tensor is freed. The copy is synchronous here, and the tests compare transcripts, which is what caught it.

## Is your PR small enough?

2 library files, +80 / -7; 2 test files, +136. Yes.

## Additional Context

**Testing**

- `test/espnet2/legacy/test_ctc_prefix_scorer.py` (new): decoding with the CTC scored on the CPU gives the same hypotheses and, to 1e-4, the same scores as scoring it on the model's device, for a CPU model and, where available, CUDA and MPS models (the MPS case ran here); scores come back on the encoder's device while the states live on the scoring device, including after `batch_select_state`; the default follows the encoder.
- `test_asr_inference.py`: the `auto` / `same` / device-name resolution, and `Speech2Text` with `ctc_scoring_device="cpu"` end to end.
- The frozen-reference tests from #6599 (221 configurations at `rtol=atol=0`), the batched parity tests, the online beam search tests and the prefix scorer reference tests pass unchanged.

**Where the rest of the step goes**, for the record: after this, a `batch_size=1` step on MPS is roughly 45 ms decoder, 10 ms `post_process`, 5 ms CTC. The decoder is a hundred-odd small kernels per step and the bookkeeping is a handful of device-to-host synchronisations; both are launch-bound too. Fusing the decoder step (`torch.compile`, CUDA graphs on CUDA) and batching the synchronisations are the next two levers. The K/V projection cache of #6624 removes arithmetic, not launches, which is why it showed nothing at `batch_size 1` and is parked.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: espnet2/bin/asr_inference.py [hunk 1]
```diff
@@ -104,6 +128,7 @@ def __init__(
         dtype: str = "float32",
         beam_size: int = 20,
         ctc_weight: float = 0.5,
+        ctc_scoring_device: str = "auto",
```
[the remark this answers is not in the document]
Comment (sw005320) [reply espnet2/bin/asr_inference.py]: Right: with the new argument before `lm_weight`, a positional call would have bound the LM weight to it. Moved to the end of the constructor signature in 49d2cef, after `max_mask_parallel`; the CLI, `inference()` and the tests pass it by keyword so nothing else moves. Also applied the other comment: the end-to-end test now asserts `len(results) == 1` instead of iterating over a possibly empty list.
